### ISSUE TITLE
After exiting the program, restore the uploaded file

### ISSUE THREAD

**Issue by xiaonaiquan** [aws/aws-sdk-cpp#2215]
### Describe the feature

I am trying write an application to continuously upload large data (multipart uploads) to Amazon's S3 Storage. However, my application needs to be able to shut down mid-transfer and pick up from where it left off the next time it's restarted.

### Use Case

From playing around a little bit with the C++ SDK, the TransferManager class provides a RetryUpload function that requires a shared pointer to the TransferHandle object that is returned upon issuing an initial UploadFile call. However, the transfer handle object will no longer be in existence if the application crashes or has to shut down mid-operation.

### Proposed Solution

resume a multipart upload using the TransferManager class

### Acknowledgements

- [ ] I may be able to implement this feature request
- [X] This feature might incur a breaking change

**Comment by jmklix** [comment]
Thanks for making this request. This is a useful feature and is something that we want to add to this sdk.

**Comment by AlexKTracerMain** [comment]
So... did you add it? Feels like the thing designed for long file uploads should be able to handle restarting after interruptions?

**Comment by xiaonaiquan** [comment]
> So... did you add it? Feels like the thing designed for long file uploads should be able to handle restarting after interruptions?

@AlexKTracerMain  yes ,Do you have any questions?

**Comment by Codebasher-Moordhan** [comment]
@xiaonaiquan @jmklix  Can you point out the API in Transfer Manager or Transfer Handle which can help resume a multi-part transfer after a system reboot, if it is implemented at all, in the AWS SDK C++? Or if there is any way to do it via present APIs gracefully & reliably?

My problem is: I have already a semi-downloaded file on disk but RAM is erased which erases the handle, so I cannot restart the download. Can/Do we have an API which can see/sees this file & can directly understand where to resume from?

**Comment by xiaonaiquan** [comment]
> [@xiaonaiquan](https://github.com/xiaonaiquan) [@jmklix](https://github.com/jmklix) Can you point out the API in Transfer Manager or Transfer Handle which can help resume a multi-part transfer after a system reboot, if it is implemented at all, in the AWS SDK C++? Or if there is any way to do it via present APIs gracefully & reliably?
> 
> My problem is: I have already a semi-downloaded file on disk but RAM is erased which erases the handle, so I cannot restart the download. Can/Do we have an API which can see/sees this file & can directly understand where to resume from?

No, you need to keep track of the upload status yourself.If you are uploading in chunks, you need to keep track of which parts have been uploaded yourself.

**Comment by kai-ion** [comment]
@Codebasher-Moordhan Thanks for the request. Resuming a multipart upload in a new process after a crash/shutdown (persisting transfer state, not just the in-memory RetryUpload/TransferHandle path) is something we want consistent across all the AWS SDKs, so it's being handled as a cross-SDK design effort rather than a C++-specific feature. It isn't being dropped; the C++ SDK will implement the agreed design once it's finalized.                                                                                                                                                                                                                        
                                                                                                                                                                                                                                                                                                                            
As a stopgap today, persist the UploadId and completed part numbers/ETags yourself and resume with the lower-level S3 multipart APIs (CreateMultipartUpload / UploadPart / ListParts / CompleteMultipartUpload), which resume across processes by design. I'm closing this ticket for now, but feel free to open one if you have anymore questions.
